File: Queue_Using_Stack.py
#Impelemnted queue using 2 stacks
#Such, that one will act as normal stack and the the other will be store the values
# of the stack in reverse order

class MyQueue:

    def __init__(self):
        self.hqueue  = []
        self.tqueue = []

    def push(self, x: int) -> None:
        self.hqueue.append(x)    

    def pop(self) -> int:
        if not self.tqueue:
            while self.hqueue:
                self.tqueue.append(self.hqueue.pop())
        x = self.tqueue.pop()
        return x
        
    def peek(self) -> int:
        if self.tqueue:
            return self.tqueue[-1]
        else:
            return self.hqueue[0]

    def empty(self) -> bool:
        if not self.hqueue and not self.tqueue:
            return True
        else:
            return False
        


# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: QueueusingStack.cs
public class MyQueue {

     private Stack<int> inStack;
    private Stack<int> outStack;

    public MyQueue() {
        inStack = new Stack<int>();
        outStack = new Stack<int>();
    }

    public void Push(int x) {
        inStack.Push(x);
    }

    public int Pop() {
        Peek(); // Ensure `outStack` is populated
        return outStack.Pop();
    }

    public int Peek() {
        if (outStack.Count == 0) {
            while (inStack.Count > 0) {
                outStack.Push(inStack.Pop());
            }
        }
        return outStack.Peek();
    }

    public bool Empty() {
        return inStack.Count == 0 && outStack.Count == 0;
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.Push(x);
 * int param_2 = obj.Pop();
 * int param_3 = obj.Peek();
 * bool param_4 = obj.Empty();
 */